File: app/src/main/java/com/bogdanorzea/popularmovies/ui/details/DescriptionTab.java
package com.bogdanorzea.popularmovies.ui.details;

import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.CursorLoader;
import android.support.v4.content.Loader;
import android.support.v4.view.ViewCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RatingBar;
import android.widget.ScrollView;
import android.widget.TextView;

import com.bogdanorzea.popularmovies.R;
import com.bogdanorzea.popularmovies.data.MoviesContract;
import com.bogdanorzea.popularmovies.model.object.Movie;
import com.bogdanorzea.popularmovies.utility.DataUtils;
import com.squareup.picasso.Picasso;

import static com.bogdanorzea.popularmovies.data.MovieMapperKt.toMovie;
import static com.bogdanorzea.popularmovies.utility.DataUtils.formatDuration;
import static com.bogdanorzea.popularmovies.utility.DataUtils.formatMoney;


public class DescriptionTab extends Fragment implements LoaderManager.LoaderCallbacks<Cursor> {
    private static final int LOADER_ID = 2;
    private int mMovieId = -1;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.layout_description, container, false);
        ScrollView scrollView = view.findViewById(R.id.scroll_view);
        ViewCompat.setNestedScrollingEnabled(scrollView, true);

        if (getArguments() != null) {
            mMovieId = getArguments().getInt("movie_id");
        }

        return view;
    }

    private void displayDescription(Cursor cursor) {
        View view = getView();
        if (view == null) {
            return;
        }

        if (cursor.moveToFirst()) {
            Movie movie = toMovie(cursor);

            // Poster
            ImageView poster = view.findViewById(R.id.poster);
            Picasso.with(getContext())
                    .load(movie.getPosterUrl())
                    .error(R.drawable.missing_cover)
                    .into(poster);

            // Release date
            ((TextView) view.findViewById(R.id.release_date))
                    .setText(DataUtils.addParenthesis(movie.getYear()));

            // Title
            ((TextView) view.findViewById(R.id.title)).setText(movie.getTitle());

            // Tagline
            ((TextView) view.findViewById(R.id.tagline)).setText(
                    DataUtils.quoteString(movie.getTagline()));

            // Overview
            ((TextView) view.findViewById(R.id.overview)).setText(movie.getOverview());

            // Score
            ((RatingBar) view.findViewById(R.id.score))
                    .setRating((float) movie.getVoteAverage() / 2);

            // Runtime
            ((TextView) view.findViewById(R.id.runtime)).setText(formatDuration(movie.getRuntime()));

            // Budget
            ((TextView) view.findViewById(R.id.budget)).setText(formatMoney(movie.getBudget()));

            // Revenue
            ((TextView) view.findViewById(R.id.revenue)).setText(formatMoney(movie.getRevenue()));

            // Genre
            ((TextView) view.findViewById(R.id.genre)).setText(movie.printGenres());
        }
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        getLoaderManager().initLoader(LOADER_ID, null, this);
        super.onActivityCreated(savedInstanceState);
    }

    @NonNull
    @Override
    public Loader<Cursor> onCreateLoader(int id, @Nullable Bundle args) {
        Uri movieUri = Uri.withAppendedPath(MoviesContract.CONTENT_URI, String.valueOf(mMovieId));

        return new CursorLoader(getActivity(), movieUri, null, null, null, null);
    }

    @Override
    public void onLoadFinished(@NonNull Loader<Cursor> loader, Cursor data) {
        displayDescription(data);
    }

    @Override
    public void onLoaderReset(@NonNull Loader<Cursor> loader) {
        loader = null;
    }
}

File: app/src/main/java/com/bogdanorzea/popularmovies/ui/details/DescriptionTab.kt
package com.bogdanorzea.popularmovies.ui.details

import android.database.Cursor
import android.net.Uri
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.LoaderManager
import android.support.v4.content.CursorLoader
import android.support.v4.content.Loader
import android.support.v4.view.ViewCompat
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.bogdanorzea.popularmovies.R
import com.bogdanorzea.popularmovies.data.MoviesContract
import com.squareup.picasso.Picasso

import com.bogdanorzea.popularmovies.data.toMovie
import com.bogdanorzea.popularmovies.utility.addParenthesis
import com.bogdanorzea.popularmovies.utility.addQuotes
import com.bogdanorzea.popularmovies.utility.toMoneyString
import com.bogdanorzea.popularmovies.utility.toTimeString
import kotlinx.android.synthetic.main.layout_description.*


class DescriptionTab : Fragment(), LoaderManager.LoaderCallbacks<Cursor> {
    companion object {
        private const val LOADER_ID = 2
    }

    private var movieId: Int = -1

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.layout_description, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        ViewCompat.setNestedScrollingEnabled(scrollView, true)

        movieId = arguments?.getInt("movie_id") ?: -1
    }

    private fun displayDescription(cursor: Cursor) {
        view ?: return

        if (cursor.moveToFirst()) {
            val movie = cursor.toMovie()

            Picasso.with(context)
                    .load(movie.getPosterUrl())
                    .error(R.drawable.missing_cover)
                    .into(poster)
            releaseDate.text = movie.getYear().addParenthesis()
            title.text = movie.title
            tagline.text = movie.tagline?.addQuotes()
            overview.text = movie.overview
            score.rating = movie.voteAverage.toFloat() / 2
            runtime.text = movie.runtime?.toTimeString()
            budget.text = movie.budget.toLong().toMoneyString()
            revenue.text = movie.revenue.toMoneyString()
            genre.text = movie.printGenres()
        }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        loaderManager.initLoader(LOADER_ID, null, this)
        super.onActivityCreated(savedInstanceState)
    }

    override fun onCreateLoader(id: Int, args: Bundle?): Loader<Cursor> {
        val movieUri = Uri.withAppendedPath(MoviesContract.CONTENT_URI, movieId.toString())

        return CursorLoader(activity!!, movieUri, null, null, null, null)
    }

    override fun onLoadFinished(loader: Loader<Cursor>, data: Cursor) {
        displayDescription(data)
    }

    override fun onLoaderReset(loader: Loader<Cursor>) {
    }
}
